packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Resources/prepareResources.ts

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2018-2021 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+import { safeDump } from 'js-yaml';
+import { cloneDeep } from 'lodash';
+import { DevWorkspaceTemplate } from '../../../../../../services/devfileApi/devfileApi';
+import { DevWorkspace } from '../../../../../../services/devfileApi/devWorkspace';
+import { DEVWORKSPACE_STORAGE_TYPE } from '../../../../../../services/devfileApi/devWorkspace/spec';
+import { generateWorkspaceName } from '../../../../../../services/helpers/generateName';
+import { DEVWORKSPACE_DEVFILE_SOURCE } from '../../../../../../services/workspace-client/devworkspace/devWorkspaceClient';
+import { DevWorkspaceResources } from '../../../../../../store/DevfileRegistries';
+
+export default function prepareResources(
+  _resources: DevWorkspaceResources,
+  factoryId: string,
+  storageType: che.WorkspaceStorageType | undefined,
+  appendSuffix: boolean,
+): DevWorkspaceResources {
+  const resources = cloneDeep(_resources);
+  const [devWorkspace, devWorkspaceTemplate] = resources;
+
+  // add the factoryId to the devfile source section
+  const { metadata } = devWorkspace;
+  if (!metadata.annotations) {
+    metadata.annotations = {};
+  }
+  metadata.annotations[DEVWORKSPACE_DEVFILE_SOURCE] = safeDump({
+    factory: { params: factoryId },
+  });
+
+  // update the DevWorkspace and DevWorkspaceTemplate names in accordance to the policy
+  if (appendSuffix == true || devWorkspace.metadata.generateName !== undefined) {
+    addSuffix(devWorkspace, devWorkspaceTemplate);
+  }
+
+  // set storage type attribute
+  if (storageType === 'ephemeral') {
+    if (!devWorkspace.spec.template.attributes) {
+      devWorkspace.spec.template.attributes = {};
+    }
+    devWorkspace.spec.template.attributes[DEVWORKSPACE_STORAGE_TYPE] = storageType;
+  } else if (devWorkspace.spec?.template?.attributes?.[DEVWORKSPACE_STORAGE_TYPE]) {
+    delete devWorkspace.spec.template.attributes[DEVWORKSPACE_STORAGE_TYPE];
+  }
+
+  return [devWorkspace, devWorkspaceTemplate];
+}
+
+function addSuffix(devWorkspace: DevWorkspace, devWorkspaceTemplate: DevWorkspaceTemplate) {
+  const suffix = generateWorkspaceName('');
```

Comment:
Could you create a separate method for generating sufixes?

Reply:
done!